src/xma/src/xmaapi/xmafilter.cpp

```diff
@@ -243,40 +196,58 @@ xma_filter_session_create(XmaFilterProperties *filter_props)
-        //Release singleton lock
-        g_xma_singleton->locked = false;
         free(filter_session->base.plugin_data);
         free(filter_session);
         delete priv1;
         return nullptr;
     }
 
-    rc = filter_session->filter_plugin->init(filter_session);
-    if (rc) {
-        xma_logmsg(XMA_ERROR_LOG, XMA_FILTER_MOD,
-                   "Initalization of filter plugin failed. Return code %d\n",
-                   rc);
-        //Release singleton lock
-        g_xma_singleton->locked = false;
-        free(filter_session->base.plugin_data);
-        free(filter_session);
-        delete priv1;
-        return nullptr;
+    //Obtain lock only for a) singleton changes & b) kernel_info changes
+    std::unique_lock<std::mutex> guard1(g_xma_singleton->m_mutex);
+    //Singleton lock acquired
+
+    if (!kernel_info->soft_kernel && !kernel_info->in_use && !kernel_info->context_opened) {
+        if (xclOpenContext(dev_handle, dev_tmp1.uuid, kernel_info->cu_index_ert, true) != 0) {
+            xma_logmsg(XMA_ERROR_LOG, XMA_FILTER_MOD, "Failed to open context to CU %s for this session\n", kernel_info->name);
+            free(filter_session->base.plugin_data);
+            free(filter_session);
+            delete priv1;
+            return nullptr;
+        }
     }
+    filter_session->base.session_id = g_xma_singleton->num_of_sessions + 1;
+    xma_logmsg(XMA_INFO_LOG, XMA_FILTER_MOD,
+                "XMA session channel_id: %d; session_id: %d\n", filter_session->base.channel_id, filter_session->base.session_id);
 
-    g_xma_singleton->num_filters++;
-    g_xma_singleton->num_of_sessions = filter_session->base.session_id;
     if (kernel_info->in_use) {
         kernel_info->is_shared = true;
+        xma_logmsg(XMA_DEBUG_LOG, XMA_FILTER_MOD,
+                   "XMA session sharing CU: %s\n", hwcfg->devices[hwcfg_dev_index].kernels[cu_index].name);
     } else {
         kernel_info->in_use = true;
+        xma_logmsg(XMA_DEBUG_LOG, XMA_FILTER_MOD,
+                   "XMA session with CU: %s\n", hwcfg->devices[hwcfg_dev_index].kernels[cu_index].name);
     }
     kernel_info->num_sessions++;
+    g_xma_singleton->num_filters++;
+    g_xma_singleton->num_of_sessions = filter_session->base.session_id;
 
-    g_xma_singleton->all_sessions.emplace(g_xma_singleton->num_of_sessions, filter_session->base);
+    g_xma_singleton->all_sessions_vec.push_back(filter_session->base);
 
     //Release singleton lock
-    g_xma_singleton->locked = false;
+    guard1.unlock();
+
+    //init can execute cu cmds as well so must be fater adding to singleton above
+    rc = filter_session->filter_plugin->init(filter_session);
+    if (rc) {
+        xma_logmsg(XMA_ERROR_LOG, XMA_FILTER_MOD,
+                   "Initalization of filter plugin failed. Return code %d\n",
+                   rc);
+        free(filter_session->base.plugin_data);
+        //free(filter_session); Added to singleton above; Keep it as checked for cu cmds
```

Comment:
Why is the filter_session not being freed or the base.plugin_data not being set to a nullptr?